### ISSUE TITLE
Unpublish with validation errors causes internal server error

### ISSUE THREAD

**Issue by PresidentNick** [wagtail/wagtail#2832]
I noticed that when saving or publishing a page with validation errors you are redirected to the page edit form with the fields that are invalid highlighted in red. If I try to unpublish a page with validation errors, I get an internal server error instead of the page form with the invalid fields highlighted. The validation check seems appropriate as that is what the changelog says https://github.com/torchbox/wagtail/blob/07ac5aa8ee5456d308789585d0375a2fbe68ede4/CHANGELOG.txt#L180 and the unpublish method calls save(). Also, this was caused by a direct change in the database (so it didn't go through the URLValidator for this URLField), it wasn't changed with the model api (that wouldn't have happened since the model api runs validator checks). So when trying to unpublish that page, it is not clear in the frontend why it is failing and causes an 500 error page instead.

**Comment by m1kola** [comment]
By the way, "Unpublish" doesn't submit form: it just points to the unpublish confirmation screen (GET request). So probably page had been saved incorrectly before unpublish method was called.

**Comment by PresidentNick** [comment]
1. version 1.4.5
2. I just put the relevant field that is making the page fail

```python
    class BaseCruiseLinePage(Page):
        image_url = models.URLField(validators=[validators.URLValidator(schemes=['https'])])
```

3. Here is the traceback

```
Traceback:
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/django/views/decorators/cache.py" in _cache_controlled
  43.             response = viewfunc(request, *args, **kw)
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/wagtail/wagtailadmin/decorators.py" in decorated_view
  22.             return view_func(request, *args, **kwargs)
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/wagtail/wagtailadmin/views/pages.py" in unpublish
  599.         page.unpublish()
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/wagtail/wagtailcore/models.py" in unpublish
  698.                 self.save()
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/django/utils/decorators.py" in inner
  145.                     return func(*args, **kwargs)
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/wagtail/wagtailcore/models.py" in save
  445.         self.full_clean()
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/wagtail/wagtailcore/models.py" in full_clean
  436.         super(Page, self).full_clean(*args, **kwargs)
File "/home/nick/.pyenv/versions/django.perx.com/lib/python2.7/site-packages/django/db/models/base.py" in full_clean
  1171.             raise ValidationError(errors)

Exception Type: ValidationError at /admin/pages/198/unpublish/
Exception Value: {'image_url': [u'Enter a valid URL.']}
```

And on your most recent comment, yeah I understand it goes to the confirmation screen with a GET request. The ValidationError is raised on the POST to that page. So in this example, `image_url` was changed to a url without the https scheme. It was written as http directly to the database (so no validation). After the POST it doesn't return to the form to show the error, just causes the server error, because of the ValidationError.

Let me know what else you may need, or if you need more clarification.

**Comment by m1kola** [comment]
Now I can reproduce the error on `master` branch with the following steps:

1. Define a model. For example:
    ```python
    class BaseCruiseLinePage(Page):
        image_url = models.URLField(validators=[validators.URLValidator(schemes=['https'])])
    
        content_panels = Page.content_panels + [
            FieldPanel('image_url')
        ]
    ```
2. Add a new page in the Wagtail admin.
3. Change a value of `image_url` directly in DB or change model definition to something like this:

    ```python
    class BaseCruiseLinePage(Page):
        image_url = models.URLField(validators=[validators.URLValidator(schemes=['http'])])
    
        content_panels = Page.content_panels + [
            FieldPanel('image_url')
        ]
    ```
    (`schemes` argument was changed from `['https']` to `['http']`)
4. Unpublish a page.

**Result:** `ValidationError`

I'm not sure that we need to handle this case. On the one hand if developer decides to change validation rules he should provide a migration for the old data, on the other hand sometimes it's not possible / necessary to migrate old date.

I've marked this issue as "Design decision needed".

Thanks @PresidentNick!

**Comment by PresidentNick** [comment]
I'm sure this can also be generalized to fields stored on the database that don't pass that field's validators, not just for URLs.

**Comment by gasman** [comment]
> I'm not sure that we need to handle this case. On the one hand if developer decides to change validation rules he should provide a migration for the old data, on the other hand sometimes it's not possible / necessary to migrate old data.

I'd agree with this - if a developer changes validation rules at any point, it's their responsibility to ensure that existing data is valid according to the new rules, so this isn't strictly a bug in Wagtail.

Having said that, we do want to keep Wagtail as developer-friendly as possible, and there isn't really any reason to perform re-validation at the point of unpublishing (and likewise for moving, where this issue has also been reported) - so if there's a nice way to disable model validation on these views without introducing lots of complexity or hacks, it would be good to do that.

**Comment by solarissmoke** [comment]
> I'd agree with this - if a developer changes validation rules at any point, it's their responsibility to ensure that existing data is valid according to the new rules, so this isn't strictly a bug in Wagtail.

Just a note that there are situations where this can happen without a developer making changes. For example if you have a nullable but non-blank foreign key field (e.g., a foreign key to `wagtailimages.Image`) which cascades to `SET_NULL` (this is a common pattern on things like image fields), and such an object is deleted.

**Comment by solarissmoke** [comment]
Fixed in #5828.
